ocaml/xapi/certs.ml

```diff
@@ -0,0 +1,414 @@
+    return {private_key; host_cert; other_certs}
+
+  let defer f = Fun.protect ~finally:f
+
+  let read_file path =
+    let ic = open_in path in
+    defer (fun () -> close_in ic) @@ fun () ->
+    really_input_string ic (in_channel_length ic)
+
+  let parse_string_using parser str =
+    let consume = Consume.Prefix in
+    parse_string ~consume parser str
+
+  let parse_file_using parser path =
+    try read_file path |> parse_string_using parser
+    with e ->
+      Error (Printf.sprintf "Can't process %s: %s" path (Printexc.to_string e))
+end
+
+(* module [Sync] provides an interface to certificates stored on
+disk. Certificates are read upon [init], potentially manipulated, and
+written back on [sync]. No other code should write these files. Writing
+back on [sync] provides a primitive form of transactions: in case of a
+fatal error, we could avoid writing a bad state to disk. This is
+currently not actively used, though *)
+
+module Sync : sig
+  type uuid = Uuidm.t
+
+  type paths = {
+      host_cert_path: string
+    ; ca_certs_dir: string
+    ; ca_crls_dir: string
+  }
+
+  type host = {
+      private_key: X509.Private_key.t
+    ; cert: X509.Certificate.t
+    ; intermediares: X509.Certificate.t list
+  }
+
+  type ca = {cert: X509.Certificate.t; uuid: uuid; mutable deleted: bool}
+
+  type cr = {name: string; cert: X509.Certificate.t; mutable deleted: bool}
+
+  (* [path] value to be used normally but other paths could be used for
+     testing *)
+  val default_paths : unit -> paths
+
+  (* initialise this module and read satte from disk *)
+  val init : paths -> (unit, [> R.msg]) result
+
+  (* write back stack in memory to disk *)
+  val sync : unit -> (unit, [> R.msg]) result
+
+  (* remove CA cert *)
+  val rm_ca : uuid -> unit -> (unit, [> R.msg]) result
+
+  (* remove CR cert *)
+  val rm_cr : string -> unit -> (unit, [> R.msg]) result
+
+  (* access CR cert *)
+  val get_cr : unit -> ((string * X509.Certificate.t) list, [> R.msg]) result
+
+  (* access CA cert *)
+  val get_ca : unit -> ((uuid * X509.Certificate.t) list, [> R.msg]) result
+
+  (* access host cert *)
+  val get_host :
+       unit
+    -> ( (X509.Private_key.t * X509.Certificate.t * X509.Certificate.t list)
+         option
+       , [> R.msg] )
+       result
+end = struct
+  type uuid = Uuidm.t
+
+  type paths = {
+      host_cert_path: string
+    ; ca_certs_dir: string
+    ; ca_crls_dir: string
+  }
+
+  type host = {
+      private_key: X509.Private_key.t
+    ; cert: X509.Certificate.t
+    ; intermediares: X509.Certificate.t list
+  }
+
+  type ca = {cert: X509.Certificate.t; uuid: uuid; mutable deleted: bool}
+
+  type cr = {name: string; cert: X509.Certificate.t; mutable deleted: bool}
+
+  type t = {
+      host: host option
+    ; mutable ca: ca list
+    ; mutable cr: cr list
+    ; paths: paths
+  }
+
+  let default_paths () =
+    {
+      host_cert_path= !Xapi_globs.server_cert_path
+    ; ca_certs_dir= "/etc/stunnel/certs"
```

Comment:
I think we're using a CA bundle rather than a CA directory (the path to which is `Stunnel.certificates_bundle_path`)

Reply:
I believe we need both and need a discussion how to handle that. But it should not be difficult to read and generate bundles as well.

Reply:
In practice both are needed. We require the names to be able to identify individual CA certificates. The bundle is needed for stunnel.

Reply:
Yes I see what's going on now. I guess we need to decide when to generate new bundles

Reply:
Right now the bundle is refreshed with a script that runs everytime CAs are synchronized in the host.